### ISSUE TITLE
Limit egress traffic on Function Apps

### ISSUE THREAD

**Issue by ryanhofdotgov** [18F/piipan#1164]
- [ ] orchestrator
- [ ] per-state
- [ ] ETL
- [ ] metrics collector
- [ ] metrics API

Use a Network Security Group on the VNet to restrict network access to the subnet, preventing any outbound traffic to Internet service tag. Be sure to specify App Service environment variable `WEBSITE_VNET_ROUTE_ALL` to force all traffic on the App Service instance through the VNet.

References:
- https://docs.microsoft.com/en-us/azure/app-service/web-sites-integrate-with-vnet
- https://docs.microsoft.com/en-us/azure/app-service/networking-features

**Comment by echappen** [comment]
@ryanhofdotgov Do we want to prevent outbound traffic over Azure's own IP addresses as well? Or just anything not an Azure IP?

**Comment by ryanhofdotgov** [comment]
@echappen Since non-trusted 3rd party sites can be hosted in Azure, it's not really allowing/denying access to Azure's IP blocks, but denying everything outside the VNet by default and only allowing communication outside the VNet to the endpoints required for the system to function.

**Comment by ryanhofdotgov** [comment]
There's some clues in[ this Slack thread](https://gsa-tts.slack.com/archives/C018P1ZJPEC/p1635466365034900?thread_ts=1635442448.007700&cid=C018P1ZJPEC), but here is the run down on how to move this forward. While this ticket scope is just about Function Apps, I'm also going to talk about Web Apps a little too.

- We have 2 app service plans, with our App Service instances and Function Apps split between them. All the Function Apps are in one (plan-apps1-dev), the web apps are in another (plan-apps2-dev). If you wanted to have tighter, more granular network level controls, you might want to split this up differently, pay the additional complexity/cost.
- Each service plan needs to have an associated subnet established in the common VNet we have created. The Function Apps already have one (vnet-core-dev/snet-apps1-dev). The web apps do not. 
- You need to add a new Network Security Group. You might be able to share this NSG for both the Function App subnet and the (future) web app subnet. On the other hand, you might be able to lock the web app one down tighter as I don't think it needs access to storage accounts. 
- For all the Function Apps to work, they need access to:
   - AAD (because they are all configured to use managed identities for authn – if you don't allow access you will get OpenID Connect errors when you make an API call)
   - Storage accounts (their own backing storage account that is used during deployment - if you don't allow access you will get Syncing.... errors during deployment, and also an error in the ETL when it can't read the bulk upload)
   - App Service (it has to be able to call some underlying implementation detail of App Service during deployment - if you don't allow access you will get Syncing... errors during deployment).
- After allowing this, you'll want to deny all outbound internet access in NSG, using the `Internet` service tag. Carefully note prioritization of the NSG rules (see screenshot).
-  The private endpoints in the VNet that we set up for the databases don't require an NSG rule
- The NSG needs to be associated with the subnet (see screenshot)
- You'll need to turn on WEBSITE_VNET_ROUTE_ALL to true. There's also a button in the portal for experimentation.
- You'll need to test every API flow (through API gateway, through the direct test scripts, through the web apps, etc.) before declaring victory that the NSG has not broken anything. Look for specific data. If it's not tested thoroughly at this point, debugging issues later will be difficult, as we don't expect internal network failures. But if this happens don't forget to see deployment log stack traces (see screenshot) as well as Activity Log for Function App.
- I would have liked to seen better granularity around enabling network access to shared Azure resources. We only want to allow network access to *our* storage accounts, the *internal* app service API it needs to call, not any Storage Account or any App Service IP in the Azure public cloud (or government cloud, as the case may be). You can ask FastTrack about this, but I don't think there's an easy way to mitigate this.
- To test that arbitrary network calls were blocked by the NSG, I temporarily modified one of the Function Apps APIs to just make a call to a server on the Internet and see if that was stalled/blocked. [Here's the code I used](https://stackoverflow.com/a/5566981) and just tried to GET `https://reqbin.com/echo/get/json` courtesy of https://reqbin.com/. If it succeeded and I saw their expected JSON Result in our API response, I knew the NSG was *not* working.
   
<img width="400" alt="Screen Shot 2021-11-02 at 1 19 36 PM" src="https://user-images.githubusercontent.com/37100189/139947623-bcbbeb0a-0583-4d70-a466-1519318f4e0b.png">
   
<img width="400" alt="Screen Shot 2021-11-02 at 1 15 42 PM" src="https://user-images.githubusercontent.com/37100189/139947629-1c7e753c-37de-4197-b66f-b30ced095a69.png">

<img width="400" alt="firewall rules" src="https://user-images.githubusercontent.com/37100189/139947651-0dfe6c61-7d7f-4695-948c-3eb3a31ba171.png">

<img width="400" alt="network error log" src="https://user-images.githubusercontent.com/37100189/139947686-d1a1efd8-ebc9-438e-9d88-dda20a3a3fdd.png">
